packages/flutter/lib/src/material/page_transitions_theme.dart

```diff
@@ -176,133 +178,162 @@ class _ZoomPageTransition extends StatefulWidget {
 
   @override
-  __ZoomPageTransitionState createState() => __ZoomPageTransitionState();
+  Widget build(BuildContext context) {
+    return _ZoomedTransitions(
+      animation: animation,
+      child: _ZoomedTransitions(
+        animation: ReverseAnimation(secondaryAnimation),
+        child: child,
+      ),
+    );
+  }
 }
 
-class __ZoomPageTransitionState extends State<_ZoomPageTransition> {
-  AnimationStatus _currentAnimationStatus;
-  AnimationStatus _lastAnimationStatus;
+class _ZoomedTransitions extends StatelessWidget {
+  const _ZoomedTransitions({
+    @required this.animation,
+    @required this.child,
+  });
+
+  final Animation<double> animation;
+  final Widget child;
 
   @override
-  void initState() {
-    super.initState();
-    widget.animation.addStatusListener((AnimationStatus animationStatus) {
-      _lastAnimationStatus = _currentAnimationStatus;
-      _currentAnimationStatus = animationStatus;
-    });
+  Widget build(BuildContext context) {
+    return DualTransitionBuilder(
+      animation: animation,
+      forwardBuilder: (
+        BuildContext context,
+        Animation<double> animation,
+        Widget child,
+      ) {
+        return _ZoomPageTransitionIn(
+          animation: animation,
+          child: child,
+        );
+      },
+      reverseBuilder: (
+        BuildContext context,
+        Animation<double> animation,
+        Widget child,
+      ) {
+        return _ZoomPageTransitionOut(
+          animation: ReverseAnimation(animation),
+          child: child,
+        );
+      },
+      child: child,
+    );
   }
+}
 
-  // This check ensures that the animation reverses the original animation if
-  // the transition were interruped midway. This prevents a disjointed
-  // experience since the reverse animation uses different fade and scaling
-  // curves.
-  bool get _transitionWasInterrupted {
-    bool wasInProgress = false;
-    bool isInProgress = false;
-
-    switch (_currentAnimationStatus) {
-      case AnimationStatus.completed:
-      case AnimationStatus.dismissed:
-        isInProgress = false;
-        break;
-      case AnimationStatus.forward:
-      case AnimationStatus.reverse:
-        isInProgress = true;
-        break;
-    }
-    switch (_lastAnimationStatus) {
-      case AnimationStatus.completed:
-      case AnimationStatus.dismissed:
-        wasInProgress = false;
-        break;
-      case AnimationStatus.forward:
-      case AnimationStatus.reverse:
-        wasInProgress = true;
-        break;
-    }
-    return wasInProgress && isInProgress;
+class _ZoomPageTransitionIn extends StatelessWidget {
+  const _ZoomPageTransitionIn({
+    this.child,
+    this.animation,
+  });
+
+  final Widget child;
+  final Animation<double> animation;
+
+  @override
+  Widget build(BuildContext context) {
+    // Only change color when the animation was completed.
+    final bool isCompleted = animation.status == AnimationStatus.completed;
+    return Container(
```

Comment:
The container will have to be wrapped in an AnimatedBuilder since you access the animation's value directly, the container needs to rebuild every time the value changes.

Reply:
Using `AnimatedBuilder` to wrap the `Container` will cause the forward transition failed. (For now idk why it'll happen).

Reply:
This one is still unresolved, as far as I can tell, there is nothing right now ensuring that the Container gets rebuild when the opacity changes.